packages/context/src/context-listener.ts

```diff
@@ -0,0 +1,53 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Binding} from './binding';
+import {ValueOrPromise} from './value-promise';
+
+/**
+ * Context event types
+ */
+export enum ContextEventType {
+  bind = 'bind',
+  unbind = 'unbind',
+}
+
+/**
+ * A function that filters bindings. It returns `true` to select a given
+ * binding.
+ */
+export type BindingFilter = (binding: Readonly<Binding<unknown>>) => boolean;
+
+/**
+ * Listeners of context bind/unbind events
+ */
+export interface ContextListener {
+  /**
+   * A filter function to match bindings
+   */
+  filter: BindingFilter;
+  /**
+   * Listen on `bind` or `unbind` and invalidate the cache
+   */
+  listen(
+    event: ContextEventType,
```

Comment:
This is not forward compatible, adding a new event type is a breaking change.

Consider an existing code compiled against an older version of context:

```ts
// Imported from Context
enum ContextEventTypeOld {
  bind = 'bind',
  unbind = 'unbind',
}

// The listener implementation
class OldListener {
  listen(event: ContextEventTypeOld, binding: Readonly<Binding<unknown>>) {
    // do some stuff
  }
}
```

Now let's say we add a new event type and release a new version of context.

```ts
export enum ContextEventType {
  bind = 'bind',
  unbind = 'unbind',
  resolved = 'resolved',
}

export interface ContextListener {
  listen(
    event: ContextEventType,
    binding: Readonly<Binding<unknown>>,
  ): ValueOrPromise<void>;
}
```

The old listener cannot be assigned to the new interface.

```ts
let listener: ContextListener;
listener = new OldListener();
```

When you concatenate all three code snippets and run `tsc`, the following error is printed:

```
x.ts:29:1 - error TS2322: Type 'OldListener' is not assignable to type 'ContextListener'.
  Types of property 'listen' are incompatible.
    Type '(event: ContextEventTypeOld, binding: Readonly<any>) => void' is not assignable to type '(event: ContextEventType, binding: Readonly<any>) => any'.
      Types of parameters 'event' and 'event' are incompatible.
        Type 'ContextEventType' is not assignable to type 'ContextEventTypeOld'.

29 listener = new OldListener();
```

Proposed solution: the interface should define an optional method for each event type. This is similar to the approach used for start/stop observers in the other pull request.

```ts
export interface ContextListener {
  onBound?(binding: Readonly<Binding<unknown>>);
  onUnbound?(binding: Readonly<Binding<unknown>>);
  // ...
}
```

Reply:
@raymondfeng I believe this comment haven't been addressed, PTAL again.

Reply:
I prefer to have one `listen` method for now to allow multiple event types. We'll revisit it as part of https://github.com/strongloop/loopback-next/pull/1972 to decide what flavors of listeners we should support.

Reply:
How do you propose to address forward-compatibility then?

The solution I am aware of is to describe the `event` argument as accepting arbitrary strings in addition to the well-known event types.

```ts
listen(
    event: ContextEventType | string,
    // ...
) {}
```

This type signature is forcing code implementing `ContextListener` to deal with event types that may be introduced in the future, which may seem like overhead right now, but it's inevitable cost of this design.